pygen/pygen_src/test/riscv_instr_cov_test.py

```diff
@@ -0,0 +1,136 @@
+        self.pad = None # Not used 
+
+class riscv_instr_cov_test():
+    """ Main class for applying the functional coverage test """
+    def __init__(self, argv):
+        self.trace_csv = []
+        self.trace = {}
+        self.csv_dir = argv.csv_dir
+        self.entry_cnt, self.total_entry_cnt, self.skipped_cnt, self.unexpected_illegal_instr_cnt = 0, 0, 0, 0
+
+    def run_phase(self):
+        self.trace_csv = os.listdir(self.csv_dir)
+        self.trace_csv = list(filter(lambda f: f.endswith('.csv'), self.trace_csv))
+        logging.info("{} CSV trace files to be processed...\n".format(len(self.trace_csv)))
+        expect_illegal_instr = False
+        for csv_file in self.trace_csv:
+            with open("{}/{}".format(self.csv_dir, csv_file)) as trace_file:
+                self.entry_cnt = 0
+                header = []
+                entry = []
+                csv_reader = csv.reader(trace_file, delimiter=',')
+                line_count = 0
+                # Get the header line
+                for row in csv_reader:
+                    if line_count == 0:
+                        header = row
+                        logging.info("Header: {}".format(header))
+                    else:
+                        entry = row
+                        if len(entry) != len(header):
+                            logging.info("Skipping malformed entry[{}]: [{}]".format(self.entry_cnt, entry))
+                            self.skipped_cnt += 1
+                        else:
+                            self.trace["csv_entry"] = row
+                            for idx in range(len(header)):
+                                if "illegal" in entry[idx]:
+                                    expect_illegal_instr = True
+                                self.trace[header[idx]] = entry[idx]
+                                if header[idx] != "pad":
+                                    logging.info("{} = {}".format(header[idx], entry[idx])) 
+                            self.post_process_trace()
+                            if self.trace["instr"] in ["li", "ret", "la"]:
+                                pass
+                            if "amo" in self.trace["instr"] or \
+                               "lr" in self.trace["instr"] or \
+                               "sc" in self.trace["instr"]:
+                                # TODO: Enable functional coverage for AMO test
+                                pass
+                            if not self.sample():
+                                if not expect_illegal_instr:
+                                    logging.error("Found unexpected illegal instr: {} [{}]".format(self.trace["instr"], entry))
+                                    self.unexpected_illegal_instr_cnt += 1
+                            # TODO: call sample here to do the functional coverage, and move assign_trace_info into it
+                            #self.sample(instruction)
```

Comment:
Are these comments necessary, since `self.sample()` is already called above?

Reply:
Oops, that's already called. I'll fix this one too.